lib/ansible/modules/cloud/amazon/ec2_import_snapshot.py

```diff
@@ -0,0 +1,250 @@
+    - ec2
+'''
+
+EXAMPLES = '''
+# Import an S3 object as a snapshot
+ec2_import_snapshot:
+  description: simple-http-server
+  format: raw
+  s3_bucket: mybucket
+  s3_key: server-image.ami
+  wait: yes
+'''
+
+RETURN = '''
+snapshot_id:
+    description: id of the created snapshot
+    returned: when snapshot is created
+    type: str
+    sample: "snap-1234abcd"
+description:
+    description: description of snapshot
+    returned: when snapshot is created
+    type: str
+    sample: "simple-http-server"
+format:
+    description: format of the disk image being imported
+    returned: when snapshot is created
+    type: str
+    sample: "raw"
+disk_image_size:
+    description: size of the disk image being imported, in bytes.
+    returned: when snapshot is created
+    type: float
+    sample: 3836739584.0
+user_bucket:
+    description: S3 bucket with the image to import
+    returned: when snapshot is created
+    type: dict
+    sample: {
+        "s3_bucket": "mybucket",
+        "s3_key": "server-image.ami"
+    }
+status:
+    description: status of the import operation
+    returned: when snapshot is created
+    type: str
+    sample: "completed"
+'''
+
+
+import time
+
+from ansible.module_utils.aws.core import AnsibleAWSModule
+from ansible.module_utils.ec2 import camel_dict_to_snake_dict, ec2_argument_spec
+
+try:
+    import botocore
+except ImportError:
+    pass
+
+
+def wait_for_import_snapshot(connection, wait_timeout, import_task_id):
+
+    params = {
+        'ImportTaskIds': [import_task_id]
+    }
+    start_time = time.time()
+    while True:
+        status = connection.describe_import_snapshot_tasks(**params)
+
+        # What are the valid status values?
+        if len(status['ImportSnapshotTasks']) > 1:
+            raise RuntimeError("Should only be 1 Import Snapshot Task with this id.")
+
+        task = status['ImportSnapshotTasks'][0]
+        if task['SnapshotTaskDetail']['Status'] in ['completed']:
+            return status
+
+        if time.time() - start_time > wait_timeout:
+            raise RuntimeError('Wait timeout exceeded (%s sec)' % wait_timeout)
+
+        time.sleep(15)
+
+
+def import_snapshot(module, connection):
+    description = module.params.get('description')
+    image_format = module.params.get('format')
+    url = module.params.get('url')
+    s3_bucket = module.params.get('s3_bucket')
+    s3_key = module.params.get('s3_key')
+    encrypted = module.params.get('encrypted')
+    kms_key_id = module.params.get('kms_key_id')
+    role_name = module.params.get('role_name')
+    wait = module.params.get('wait')
+    wait_timeout = module.params.get('wait_timeout')
+
+    if not url and not (s3_bucket and s3_key):
+        module.fail_json(msg='url or s3_bucket and s3key are required.')
+    if all([url, s3_bucket, s3_key]):
+        module.fail_json(msg='url or s3_bucket and s3key are required, not all 3.')
+    if not image_format:
+        module.fail_json(msg='format is required.')
```

Comment:
It's better to use the standard parameter validation instead of rolling your own. For image_format, just set `required=True` in the argument spec. The rest could be expressed as `mutually_exclusive=[['s3_bucket', 'url']], required_one_of=[['s3_bucket', 'url']], required_together=[['s3_bucket', 's3_key']]` in the module instantiation.